File: Example/ObjcExample/TableView/TableViewController.m
//  TableViewController.m
//  ObjCSkeletonDemo
//
//  Created by Mikhail Zoline on 8/13/20.
//  Copyright © 2020 MZ. All rights reserved.
//

@import SkeletonView;
#import "ObjcExample-Swift.h"
#import "TableViewCell.h"
#import "TableViewController.h"

@interface TableViewController ()

@property (weak, nonatomic) IBOutlet UITableView *tableView;
@property (weak, nonatomic) IBOutlet UIImageView *avatarImage;
@property (weak, nonatomic) IBOutlet UIView *colorSelectedView;
@property (weak, nonatomic) IBOutlet UISwitch *switchAnimated;
@property (weak, nonatomic) IBOutlet UISegmentedControl *skeletonTypeSelector;
@property (weak, nonatomic) IBOutlet UIButton *showOrHideSkeletonButton;

@property (weak, nonatomic) IBOutlet UILabel *transitionDurationLabel;
@property (weak, nonatomic) IBOutlet UIStepper *transitionDurationStepper;

@property (nonatomic, readonly, assign) BOOL type;
@end

@implementation TableViewController

- (void)configureTableView {
    _tableView.isSkeletonable = true;
    _tableView.rowHeight = UITableViewAutomaticDimension;
    _tableView.estimatedRowHeight = 120.0;
    _tableView.sectionHeaderHeight = UITableViewAutomaticDimension;
    _tableView.estimatedSectionHeaderHeight = 20.0;
    _tableView.sectionFooterHeight = UITableViewAutomaticDimension;
    _tableView.estimatedSectionFooterHeight = 20.0;
    
    [_tableView registerClass: HeaderFooterSection.class forHeaderFooterViewReuseIdentifier:@"FooterIdentifier"];
    [_tableView registerClass: HeaderFooterSection.class forHeaderFooterViewReuseIdentifier:@"HeaderIdentifier"];
}

- (void)configureColorSelectedView {
    _colorSelectedView.layer.cornerRadius = 5;
    _colorSelectedView.layer.masksToBounds = true;
    _colorSelectedView.backgroundColor = SkeletonAppearance.shared.tintColor;
    _colorSelectedView.tintColor = [Colors color:16];
}

- (void)configureAvatarImage {
    _avatarImage.layer.cornerRadius = self.avatarImage.frame.size.width/2;
    _avatarImage.layer.masksToBounds = true;
}

- (void)viewDidLoad {
    [super viewDidLoad];
    _transitionDurationStepper.value = 0.25;
    [self configureTableView];
    [self configureColorSelectedView];
    [self configureAvatarImage];
    [self showSkeleton];
}

- (IBAction)changeAnimated:(id)sender {
    if (_switchAnimated.isOn) {
        [self.view startSkeletonAnimation:NULL];
    } else {
        [self.view stopSkeletonAnimation];
    }
}

- (IBAction)btnChangeColorTouchUpInside:(id)sender {
    [self showAlertPicker];
}

- (IBAction)transitionDurationStepperAction:(id)sender {
    _transitionDurationLabel.text = [NSString stringWithFormat:@"Fade duration: %.2f s", _transitionDurationStepper.value];
    if (self.view.isSkeletonActive) { [self showSkeleton]; }
}

- (IBAction)changeSkeletonType:(id)sender {
    [self refreshSkeleton];
}

- (IBAction)showOrHideSkeleton:(id)sender {
    [_showOrHideSkeletonButton setTitle: self.view.isSkeletonActive ? @"Show skeleton" : @"Hide skeleton" forState:UIControlStateNormal];
    self.view.isSkeletonActive ? [self hideSkeleton] : [self showSkeleton];
}


- (BOOL)type {
    return _skeletonTypeSelector.selectedSegmentIndex == 0 ? SkeletonType.solid: SkeletonType.gradient;
}

- (void)showSkeleton {
    [self refreshSkeleton];
}

- (void)hideSkeleton {
    [self.view hideSkeletonWithReloadDataAfter:
     true transition:
     [[SkeletonTransitionStyle alloc]init:
      _transitionDurationStepper.value]];
}

- (void)refreshSkeleton {
    if (self.view.isSkeletonActive) {
        SkeletonTransitionStyle *style = [[SkeletonTransitionStyle alloc]init: 0];
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.view hideSkeletonWithReloadDataAfter:true transition: style];
        });
    }
    
    [NSThread sleepForTimeInterval:0.1f];
    
    dispatch_async(dispatch_get_main_queue(), ^{
        if ( self.type == SkeletonType.gradient) {
            [self showGradientSkeleton];
        } else {
            [self showSolidSkeleton ];
        }
    });
}

- (void)showSolidSkeleton {
    if (_switchAnimated.isOn) {
        [self.view showAnimatedSkeletonUsingColor:
         _colorSelectedView.backgroundColor animation:
         NULL transition:
         [[SkeletonTransitionStyle alloc]init:
          _transitionDurationStepper.value]];
    } else {
        [self.view showSkeletonUsingColor:
         _colorSelectedView.backgroundColor transition:
         [[SkeletonTransitionStyle alloc]init:
          _transitionDurationStepper.value]];
    }
}

- (void)showGradientSkeleton {
    SkeletonGradient* gradient = [[SkeletonGradient alloc] initWithBaseColor:_colorSelectedView.backgroundColor secondaryColor:_tableView.tintColor];
    
    if (_switchAnimated.isOn) {
        [self.view showAnimatedGradientSkeletonUsingGradient:
         gradient animation:
         NULL transition:
         [[SkeletonTransitionStyle alloc]init:
          _transitionDurationStepper.value]];
    } else {
        [self.view showGradientSkeletonUsingGradient:
         gradient transition:
         [[SkeletonTransitionStyle alloc]init:
          _transitionDurationStepper.value]];
    }
}

+ (NSArray *)lorem
{
    static NSArray *_lorem;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        _lorem = @[@"Lorem ipsum", @"dolor sit", @"er elit", @"lamet", @"consectetaur", @"cillium", @"adipisicing", @"pecu sed", @"do eiusmod", @"tempor", @"incididunt", @"ut labore", @"et dolore", @"magna aliqua" ];
    });
    return _lorem;
}

- (void)showAlertPicker {
    UIAlertController* alertView =
    [UIAlertController alertControllerWithTitle:
     @"Select color" message:@"\n\n\n\n\n\n" preferredStyle:
     UIAlertControllerStyleAlert];
    
    UIPickerView* pickerView =
    [[UIPickerView alloc] initWithFrame:
     CGRectMake(0.0, 50.0, 260.0, 120.0)];
    [pickerView setDataSource:self];
    [pickerView setDelegate:self];
    [alertView.view addSubview:pickerView];
    
    UIAlertAction* action =
    [UIAlertAction actionWithTitle:
     @"Ok" style:
     UIAlertActionStyleDefault handler:
     ^(UIAlertAction * _Nonnull action) {
        __weak TableViewController* weakSelf = self;
        __weak UIPickerView* weakPicker = pickerView;
        long row = [weakPicker selectedRowInComponent:0];
        weakSelf.colorSelectedView.backgroundColor = [Colors color:row];
        [weakSelf refreshSkeleton];
    }];
    [alertView addAction:action];
    
    UIAlertAction* cancelAction =
    [UIAlertAction actionWithTitle:
     @"Cancel" style:
     UIAlertActionStyleCancel handler:
     NULL];
    [alertView addAction:cancelAction];
    [self presentViewController:
     alertView animated:
     false completion:
     NULL];
}

// MARK: - UIPickerViewDelegate, UIPickerViewDataSource
- (NSInteger)numberOfComponentsInPickerView:(UIPickerView *)pickerView {
    return 1;
}

- (NSInteger)pickerView:(UIPickerView *)pickerView numberOfRowsInComponent:(NSInteger)component {
    return [Colors count];
}

- (NSString *)pickerView:(UIPickerView *)pickerView titleForRow:(NSInteger)row forComponent:(NSInteger)component {
    return [Colors title:row] ;
}

// MARK: - SkeletonTableViewDataSource

- (NSInteger)numSectionsIn:(UITableView *)collectionSkeletonView {
    return 1;
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return 10;
}

- (NSInteger)collectionSkeletonView:(UITableView *)skeletonView numberOfRowsInSection:(NSInteger)section {
    return 10;
}

- (NSString *)collectionSkeletonView:(UITableView *)skeletonView cellIdentifierForRowAt:(NSIndexPath *)indexPath {
    return @"CellIdentifier";
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    TableViewCell *cell = (TableViewCell *)[tableView dequeueReusableCellWithIdentifier:@"CellIdentifier"];
    cell.label1.text = TableViewController.lorem[indexPath.row];
    
    return cell;
}
// MARK: - SkeletonTableViewDelegate
- (NSString *)collectionSkeletonView:(UITableView *)skeletonView identifierForHeaderInSection:(NSInteger)section {
    return @"HeaderIdentifier";
}

- (UIView *)tableView:(UITableView *)tableView viewForHeaderInSection:(NSInteger)section {
    HeaderFooterSection *header = (HeaderFooterSection*)[tableView dequeueReusableHeaderFooterViewWithIdentifier:@"HeaderIdentifier"];
    header.textLabel.text = TableViewController.lorem.firstObject;
    return header;
}

- (NSString *)collectionSkeletonView:(UITableView *)skeletonView identifierForFooterInSection:(NSInteger)section {
    return @"FooterIdentifier";
}

- (UIView *)tableView:(UITableView *)tableView viewForFooterInSection:(NSInteger)section {
    HeaderFooterSection *footer = (HeaderFooterSection*)[tableView dequeueReusableHeaderFooterViewWithIdentifier:@"FooterIdentifier"];
    footer.textLabel.text = TableViewController.lorem.lastObject;
    return footer;
}

@end

File: Example/SwiftExample/TableView/TableViewController.swift
//
//  ViewController.swift
//  SkeletonViewExample
//
//  Created by Juanpe Catalán on 02/11/2017.
//  Copyright © 2017 SkeletonView. All rights reserved.
//

import UIKit
import SkeletonView

class TableViewController: UIViewController {
    @IBOutlet weak var tableview: UITableView! {
        didSet {
            tableview.rowHeight = UITableView.automaticDimension
            tableview.sectionHeaderHeight = UITableView.automaticDimension
            tableview.sectionFooterHeight = UITableView.automaticDimension
            tableview.estimatedRowHeight = 120.0
            tableview.estimatedSectionFooterHeight = 20.0
            tableview.estimatedSectionHeaderHeight = 20.0
            tableview.register(HeaderFooterSection.self, forHeaderFooterViewReuseIdentifier: "HeaderIdentifier")
            tableview.register(HeaderFooterSection.self, forHeaderFooterViewReuseIdentifier: "FooterIdentifier")
        }
    }
    
    @IBOutlet weak var avatarImage: UIImageView! {
        didSet {
            avatarImage.layer.cornerRadius = avatarImage.frame.width/2
            avatarImage.layer.masksToBounds = true
        }
    }
    
    @IBOutlet weak var colorSelectedView: UIView! {
        didSet {
            colorSelectedView.layer.cornerRadius = 5
            colorSelectedView.layer.masksToBounds = true
            colorSelectedView.backgroundColor = SkeletonAppearance.shared.tintColor
        }
    }

    @IBOutlet weak var switchAnimated: UISwitch!
    @IBOutlet weak var skeletonTypeSelector: UISegmentedControl!
    @IBOutlet weak var showOrHideSkeletonButton: UIButton!
    @IBOutlet weak var transitionDurationLabel: UILabel!
    @IBOutlet weak var transitionDurationStepper: UIStepper!
    
    var type: SkeletonType {
        return SkeletonType(solidOrGradient: skeletonTypeSelector.selectedSegmentIndex == 0 ?  SkeletonType.solid : SkeletonType.gradient)
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        transitionDurationStepper.value = 0.25
        self.view.showAnimatedGradientSkeleton()
    }
    
    override func viewDidAppear(_ animated: Bool) {
        super.viewDidAppear(animated)
    }
    //(CALayer) -> CAAnimation
    class var defaultAnim: SkeletonLayerAnimation {
        get {
            let animBuilder  = GradientDirection(direction: GradientDirection.topLeftBottomRight)
            return animBuilder.slidingAnimation()
        }
    }

    @IBAction func changeAnimated(_ sender: Any) {
        if switchAnimated.isOn {
            view.startSkeletonAnimation(TableViewController.defaultAnim)
        } else {
            view.stopSkeletonAnimation()
        }
    }
    
    @IBAction func changeSkeletonType(_ sender: Any) {
        if self.view.isSkeletonActive {
            self.hideSkeleton()
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
                self.showSkeleton()
            }
        }
    }
    
    @IBAction func btnChangeColorTouchUpInside(_ sender: Any) {
        showAlertPicker()
    }
    
    @IBAction func showOrHideSkeleton(_ sender: Any) {
        showOrHideSkeletonButton.setTitle((view.isSkeletonActive ? "Show skeleton" : "Hide skeleton"), for: .normal)
        view.isSkeletonActive ? hideSkeleton() : showSkeleton()
    }
    
    @IBAction func transitionDurationStepperAction(_ sender: Any) {
        transitionDurationLabel.text = NSString(format:"Fade duration: %.2f s", transitionDurationStepper.value) as String
        if view.isSkeletonActive { showSkeleton() }
    }
    
    func showSkeleton() {
        if type.isGardient { showGradientSkeleton() }
        else { showSolidSkeleton() }
    }

    func hideSkeleton() {
        view.hideSkeleton(transition: SkeletonTransitionStyle(transitionDurationStepper.value))
    }
    
    func showSolidSkeleton() {
        if switchAnimated.isOn {
            view.showAnimatedSkeleton(usingColor: colorSelectedView.backgroundColor!, transition: SkeletonTransitionStyle(transitionDurationStepper.value))
        } else {
            view.showSkeleton(usingColor: colorSelectedView.backgroundColor!, transition: SkeletonTransitionStyle(transitionDurationStepper.value))
        }
    }
    
    func showGradientSkeleton() {
        let gradient = SkeletonGradient(baseColor: colorSelectedView.backgroundColor!, secondaryColor: colorSelectedView.tintColor)
        if switchAnimated.isOn {
            view.showAnimatedGradientSkeleton(usingGradient: gradient, transition: SkeletonTransitionStyle(transitionDurationStepper.value))
        } else {
            view.showGradientSkeleton(usingGradient: gradient, transition: SkeletonTransitionStyle(transitionDurationStepper.value))
        }
    }
    
    func showAlertPicker() {
        let alertView = UIAlertController(title: "Select color", message: "\n\n\n\n\n\n", preferredStyle: .alert)
        
        let pickerView = UIPickerView(frame: CGRect(x: 0, y: 50, width: 260, height: 115))
        pickerView.dataSource = self
        pickerView.delegate = self
        
        alertView.view.addSubview(pickerView)
        
        let action = UIAlertAction(title: "OK", style: .default) { [unowned pickerView, unowned self] _ in
            let row = pickerView.selectedRow(inComponent: 0)
            self.colorSelectedView.backgroundColor = Colors.color(row)
            self.colorSelectedView.tintColor = Colors.color(row+1 < Colors.count ?  row+1 : 0 )
            if self.view.isSkeletonActive {
                self.hideSkeleton()
                DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
                    self.showSkeleton()
                }
            }
        }
        alertView.addAction(action)
        
        let cancelAction = UIAlertAction(title: "Cancel", style: .cancel, handler: nil)
        alertView.addAction(cancelAction)
        
        present(alertView, animated: false, completion: {
            pickerView.frame.size.width = alertView.view.frame.size.width
        })
    }
    
    static var lorem = String("Lorem ipsum dolor sit er elit lamet, consectetaur cillium adipisicing pecu, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.").components(separatedBy: .whitespaces)
    static var currIdx = 0
}

extension TableViewController: UIPickerViewDelegate, UIPickerViewDataSource {
    func numberOfComponents(in pickerView: UIPickerView) -> Int {
        return 1
    }
    
    func pickerView(_ pickerView: UIPickerView, numberOfRowsInComponent component: Int) -> Int {
       return Colors.count
    }
    
    func pickerView(_ pickerView: UIPickerView, titleForRow row: Int, forComponent component: Int) -> String? {
        return Colors.title(row)
    }
}

extension TableViewController: SkeletonTableViewDataSource {

    @objc func numSections(in collectionSkeletonView: UITableView) -> Int {
        return 1
    }
    @objc func collectionSkeletonView(_ skeletonView: UITableView, numberOfRowsInSection section: Int) -> Int {
      return 10
    }
    @objc func collectionSkeletonView(_ skeletonView: UITableView, cellIdentifierForRowAt indexPath: IndexPath) -> ReusableCellIdentifier {
        return "CellIdentifier"
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return 10
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell = tableView.dequeueReusableCell(withIdentifier: "CellIdentifier", for: indexPath) as! Cell
        cell.label1.text = TableViewController.lorem[indexPath.row]
        if let header = tableView.headerView(forSection: indexPath.section) {
            (header as! HeaderFooterSection).titleLabel.text = TableViewController.lorem[indexPath.row]
        }
        if let header = tableView.headerView(forSection: indexPath.section) {
            (header as! HeaderFooterSection).titleLabel.text = TableViewController.lorem[indexPath.row > 1 ? indexPath.row - 2 : indexPath.row ]
        }
        if let footer = tableView.footerView(forSection: indexPath.section) {
            (footer as! HeaderFooterSection).titleLabel.text = TableViewController.lorem[indexPath.row < TableViewController.lorem.count ? indexPath.row + 1 : indexPath.row ]
        }
        return cell
    }
}

extension TableViewController: SkeletonTableViewDelegate {
    func collectionSkeletonView(_ skeletonView: UITableView, identifierForHeaderInSection section: Int) -> ReusableHeaderFooterIdentifier? {
        return "HeaderIdentifier"
    }

    func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
        let header = tableView
            .dequeueReusableHeaderFooterView(withIdentifier: "HeaderIdentifier") as! HeaderFooterSection
        header.titleLabel.text =  TableViewController.lorem[0]
        return header
    }

    func collectionSkeletonView(_ skeletonView: UITableView, identifierForFooterInSection section: Int) -> ReusableHeaderFooterIdentifier? {
        return "FooterIdentifier"
    }

    func tableView(_ tableView: UITableView, viewForFooterInSection section: Int) -> UIView? {
        let footer = tableView
            .dequeueReusableHeaderFooterView(withIdentifier: "FooterIdentifier") as! HeaderFooterSection
        footer.titleLabel.text = TableViewController.lorem[2]
        return footer
    }
}
